operators/pkg/tenant-controller/tenant_controller.go

```diff
@@ -307,6 +311,86 @@ func (r *TenantReconciler) checkValidWorkspaces(ctx context.Context, tn *crownla
 	return tenantExistingWorkspaces, workspaces, err
 }
 
+// deleteClusterNamespace deletes the namespace for the tenant, if it fails then it returns an error.
+func (r *TenantReconciler) deleteClusterNamespace(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string) (err error) {
+	ns := v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: nsName}}
+
+	nsErr := utils.EnforceObjectAbsence(ctx, r.Client, &ns, "personal namespace")
+
+	if nsErr != nil {
+		klog.Errorf("Error when deleting namespace of tenant %s -> %s", tn.Name, nsErr)
+	}
+
+	return nsErr
+}
+
+// enforceNamespaceKeepAliveOrDelete deletes the namespace if it should be deleted.
+func (r *TenantReconciler) enforceNamespaceKeepAliveOrDelete(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string, tenantWorkspaceKeepAlive time.Duration) (keepNsOpen bool, err error) {
+	// We check to see if last login was more than tenantWorkspaceKeepAlive in the past:
+	// if so, temporarily delete the namespace. We assume that a lastLogin of 0 occurs when a user is first created
+
+	// Calculate time elapsed since lastLogin (now minus lastLogin in seconds)
+	sPassed := time.Since(tn.Spec.LastLogin.Time)
+
+	klog.Infof("Last login of tenant %s was %s ago", tn.Name, sPassed)
+
+	// Attempt to get instances in current namespace
+	list := &crownlabsv1alpha2.InstanceList{}
+
+	if err := r.List(context.Background(), list, client.InNamespace(nsName)); err != nil {
+		return true, err
+	}
+
+	if sPassed > tenantWorkspaceKeepAlive { // seconds
+		klog.Infof("Over %s elapsed since last login of tenant %s: attempting to delete tenant namespace if not already deleted", tenantWorkspaceKeepAlive, tn.Name)
+		if len(list.Items) == 0 {
+			klog.Infof("No instances in %s: workspace can be deleted", nsName)
+			keepNsOpen = false
+		} else {
+			klog.Infof("Instances in namespace %s. Namespace will not be deleted", nsName)
+		}
+	} else {
+		klog.Infof("Under %s (limit) elapsed since last login of tenant %s: namespace is left as-is", tenantWorkspaceKeepAlive, tn.Name)
+	}
+
+	return keepNsOpen, nil
+}
+
+// Deletes namespace or updates the cluster resources.
+func (r *TenantReconciler) enforceClusterResources(ctx context.Context, tn *crownlabsv1alpha2.Tenant, nsName string, keepNsOpen bool) (nsOk bool, err error) {
+	nsOk = false // nsOk must be initialized for later use
```

Comment:
I would suggest to completely drop the nsOk part for the sake of simplicity. just stick to the fact that no error means OK, while in case of error something went wrong.

Reply:
I think this would require refactoring of the `createOrUpdateClusterResources` function, which is part of a different functionality. For this reason I would suggest to leave in nsOk for consistency until that other refactoring step is completed by the project maintainers